Fix parser bug between regex and div operator

wateret opened the pull request:
For Lexer alone, it is hard to find out whether a token that starts with `/` is a division operator or a regex literal. Parser should give its context information to Lexer that it is expecting which. 

In my implementation, Lexer expects division operator by default.
When Parser sees token `/` or `/=` in `parse_primary_expression()`, the token is re-lexed as a regex literal. Seems work good.

This will resolve #652 and #655.

galpeter commented:
Few points:
* When modifying method signatures, please update also the relevant method comments. For example: the comment on `lexer_parse_token` was not updated. (Also the parameter comment is missing from the method)
* You mentioned the two issues (#652 and #655), those are good test cases, we should add them into our `tests/jerry/` directory
* Partially related: when I was trying out the PR I've found out that the parsing of the following js code is incorrect:
```js
1 / 2
/=a/;
```
The returned error message is: `ERROR: Ln 2, Col 5: Unknown token ;`. However this error also occurs without this PR

wateret commented:
@galpeter 
Thank you for your feedback. I made some changes:
* Updated comments as commented.
* Added test cases of #652 and #655. Added additional tricky test cases in `tests/jerry/regexp-literal.js`
* I'm not 100% sure about this, but it seems like no semicolon is inserted after first line(`1 / 2`) in your code. (Major JS engines like `v8` does not insert a semicolon.)
 * If so, the next token should be parsed as `/=`. This PR doesn't affect it.
 * If not, this PR will successfully parse the next token as `/=a/`.
 * Whichever, I think this issue is more about *automatic semicolon insertion*.

sand1k commented:
This resolves #563 too.

sand1k commented on jerry-core/parser/js/parser.cpp:
```diff
@@ -830,6 +830,14 @@ parse_primary_expression (void)
+      // must be a regexp literal so re-lex token as a regexp literal
+      lexer_seek (tok.loc);
+      skip_token (true);
+      /* FALLTHRU */
```
I would prefer to add a function:
```
static void
rescan_regexp_token ()
{
  lexer_seek (tok.loc);
  tok = lexer_next_token (true);
}
```
and call it here.  This would allow removing default argument from `skip_token` and make the code more explicit.

wateret commented:
@sand1k 
Thanks, I totally agree with you.
Updated as commented and added a test case for #563.

sand1k commented on jerry-core/parser/js/lexer.cpp:
```diff
@@ -1218,10 +1218,15 @@ lexer_parse_comment (void)
  */
 static token
-lexer_parse_token (void)
+lexer_parse_token (bool maybe_regexp = false) /**< read '/' as regexp? */
```
Default argument value is not required here. Every call site of `lexer_parse_token` passes an argument.

sand1k commented on jerry-core/parser/js/lexer.h:
```diff
@@ -171,7 +171,7 @@ typedef struct
 void lexer_init (const jerry_api_char_t *, size_t, bool);
 
-token lexer_next_token (void);
+token lexer_next_token (bool = false);
```
Could you please remove default argument value and use explicit argument value at every call site. We've agreed to use C++ features as little as possible in Jerry.

sand1k commented on jerry-core/parser/js/parser.cpp:
```diff
@@ -99,6 +99,13 @@ skip_token (void)
 }
 
 static void
+rescan_regexp_token (void)
```
Please add comment with function description.

sand1k commented on jerry-core/parser/js/parser.cpp:
```diff
@@ -99,6 +99,13 @@ skip_token (void)
+{
+  lexer_seek (tok.loc);
+  tok = lexer_next_token (true);
+}
```
/* rescan_regexp_token */

sand1k commented:
@wateret, please, fix minor issues.
As for the rest, LGTM.

sand1k commented on jerry-core/parser/js/parser.cpp:
```diff
@@ -252,7 +262,7 @@ jsp_find_next_token_before_the_locus (token_type token_to_find, /**< token to se
         {
           lexer_seek (end_loc);
-          tok = lexer_next_token ();
+          tok =  lexer_next_token (false);
```
Extra space here and below.

wateret commented:
@sand1k Sorry for the extra spaces. I fixed them.
I think I need one more review to merge this?

egavrin commented:
@galpeter could you please check this PR?

galpeter commented:
lgtm

egavrin commented:
@wateret `make push`
